python/src/uagents/context.py

```diff
@@ -585,136 +479,180 @@ async def send_raw_exchange_envelope(
 
-            except Exception as ex:
-                log(
-                    logger,
-                    logging.WARNING,
-                    f"Failed to send message to {destination} @ {endpoint}: {ex}",
-                )
+        # Create awaitable future for MsgStatus and sync response
+        fut = asyncio.Future()
 
-        log(logger, logging.ERROR, f"Failed to deliver message to {destination}")
+        self._queue_envelope(env, endpoints, fut, sync)
 
-        return MsgStatus(
-            status=DeliveryStatus.FAILED,
-            detail="Message delivery failed",
-            destination=destination,
-            endpoint="",
-        )
+        try:
+            result = await asyncio.wait_for(fut, timeout)
+        except asyncio.TimeoutError:
+            log(self.logger, logging.ERROR, "Timeout waiting for dispense response")
+            return MsgStatus(
+                status=DeliveryStatus.FAILED,
+                detail="Timeout waiting for response",
+                destination=destination,
+                endpoint="",
+            )
 
-    @staticmethod
-    async def dispatch_sync_response_envelope(env: Envelope) -> MsgStatus:
-        await dispatcher.dispatch(
-            env.sender,
-            env.target,
-            env.schema_digest,
-            env.decode_payload(),
-            env.session,
-        )
-        return MsgStatus(
-            status=DeliveryStatus.DELIVERED,
-            detail="Sync message successfully delivered via HTTP",
-            destination=env.target,
-            endpoint="",
-        )
+        if isinstance(result, Envelope):
+            return await dispatch_sync_response_envelope(result)
+
+        return result
+
+    def _queue_envelope(
+        self,
+        envelope: Envelope,
+        endpoints: List[str],
+        response_future: asyncio.Future,
+        sync: bool = False,
+    ):
+        """
+        Queue an envelope for processing.
+
+        Args:
+            envelope (Envelope): The envelope to queue.
+        """
+        self._dispenser.add_envelope(envelope, endpoints, response_future, sync)
 
     async def send_wallet_message(
         self,
         destination: str,
         text: str,
         msg_type: int = 1,
-    ):
+    ):  # TODO: restructure wallet messaging
         if self._wallet_messaging_client is not None:
             await self._wallet_messaging_client.send(destination, text, msg_type)
         else:
-            self.logger.warning("Cannot send wallet message: no client available")
+            log(
+                self.logger,
+                logging.WARNING,
+                "Cannot send wallet message: no client available",
+            )
 
 
-async def send_sync_message(
-    destination: str,
-    message: Model,
-    response_type: Type[Model] = None,
-    sender: Identity = None,
-    resolver: Resolver = None,
-    timeout: int = 30,
-) -> Union[Model, JsonStr, MsgStatus]:
+class ExternalContext(InternalContext):
```

Comment:
I don't have a good suggestion yet, but I feel we could improve on the naming and description of these contexts. It should be clear what "Internal", "External", "reactive" and "proactive" etc means to both new developers as well as those that already worked with the previous solitary context

Reply:
You are absolutely right, the naming is essential but everything that is changed in this refactor is not visible to the developer. We purposely use the new Context Interface class to specifically design the usable aspects of the Context. Nothing that is changed in this PR should affect the user at all.